students/etwum/lesson04/mailroom_part2.py

```diff
@@ -0,0 +1,153 @@
+                          "avg donation": new_donation_amount})
+        lst_Donors.extend(new_donor)
+        print('-----------------------------------------------------')
+        for x in lst_Donors:
+            if x["donor name"] == donor_full_name:
+                email = "Dear {a},\n\nThank you for your generous donations of ${b:.2f} to our charity.\n".format(
+                    a=x["donor name"],b=x["total donations"])
+                print(email)
+    print('-----------------------------------------------------')
+
+
+def sort_list(lst_Donors):
+    # used to sort the donor list by the total donations column
+
+    return lst_Donors["total donations"]
+
+
+def create_report():
+    # creates a report of the the donors
+    # headers used in table
+    print()
+    lst_Header = [["Donor Name", "| Total Donation(s)", "| # of Donations", "| Avg Donation"]]
+
+    for x in lst_Header:
+        print('{:<25}{:<20}{:<17}{:<15}'.format(*x))
+    print("----------------------------------------------------------------------------")
+    for x in sorted(lst_Donors,key=sort_list, reverse= True):
+        print('{:<25} $ {:<20}{:^14} $ {:<15}'.format(*x.values()))
+
+
+def send_letter_all():
+    print('\n----------------------------------------------')
+    file = ()
+    # creates a letter to every donor by writing them to separate text files
+    for x in lst_Donors:
+        # opens or creates a new text file for writing based on the donor name
+        file = open('{a}.txt'.format(a = x['donor name']), 'w')
+
+        # writes a thank you letter to the text file
+        file.write('Dear {a},\n\nThank you for your generous donation(s) of ${b:.2f} to our charity.'.format(
+            a=x['donor name'], b=x['total donations']))
+        print('Letter sent to  {a}'.format(a=x['donor name']))
+
+    #closes and saves the file
+    file.close()
+    print('----------------------------------------------\n')
+
+
+print("Welcome to the Charity Mail Room")
+print("------------------------------------------------------------------------")
+
+str_choice = 0
+# ****Input/Output****
+while str_choice != 4:
+    # Option menu
+
+    print("""
+    Menu of Options
+    1) Send Thank You Note
+    2) Create Report
+    3) Send Letter to Everyone
+    4) Exit Program
+    """)
+
+    # dictionary used to run different functions of the program
+    User_Selection = {1: send_thank_you, 2: create_report, 3: send_letter_all}
+
+    # try/except block to make sure the user inputs a valid option
+    try:
+        str_choice = int(input("Which option would you like to perform? Input a number [1 to 4] "))
+        if str_choice in (1,2,3):
+            # runs a function selection from the dictionary
+            User_Selection[str_choice]()
+        elif str_choice == 4:
+            print("\nExiting Program")
+        else:
+            raise Exception
+    except Exception:
+        print("Please input a valid option: 1, 2, 3, or 4")
```

Comment:
I would put this into `main()` function and call `main()` in `if __name__ == __main__`